### PR TITLE
fix: fix too early loop exit

### LINKED ISSUE #56: Test 6.3.9
Test 6.3.9 has some bugs: (See https://github.com/oasis-tcs/csaf/pull/583/files for test files)
- valid files are considered invalid: e.g. oasis_csaf_tc-csaf_2_0-2021-6-3-09-12.json
    ```
	{
	  "tests": [
		{
		  "isValid": true,
		  "errors": [],
		  "warnings": [],
		  "infos": [
			{
			  "instancePath": "/product_tree/branches/0/branches/0/branches/0/branches/1",
			  "message": "missing ancestor with category vendor"
			}
		  ],
		  "name": "informativeTest_6_3_9"
		}
	  ],
	  "isValid": true
	}
   ```
- invalid files do no show all errors: e.g. oasis_csaf_tc-csaf_2_0-2021-6-3-09-03.json reports only on instead of 4 errors (for 2 elements both defined error messages)
  ```
  	{
	  "tests": [
		{
		  "isValid": true,
		  "errors": [],
		  "warnings": [],
		  "infos": [
			{
			  "instancePath": "/product_tree/branches/0/branches/0/branches/0",
			  "message": "missing ancestor with category vendor"
			}
		  ],
		  "name": "informativeTest_6_3_9"
		}
	  ],
	  "isValid": true
	}
    ```

### PR DESCRIPTION
**Pull request by domachine**
Fix #56

### CODE REVIEW

File: lib/informativeTests/informativeTest_6_3_9.js [hunk 1]
```diff
@@ -81,7 +81,7 @@ export default function informativeTest_6_3_9(doc) {
                 instancePath: `${prefix}${branchIndex}`,
                 message: `missing ancestor with category ${index.category}`,
               })
-              return
+              break
```
Comment (tschmidtb51) [inline lib/informativeTests/informativeTest_6_3_9.js:84]: This leads to the point, where only the first missing category is reported. I guess, we should report them all at once.
```suggestion
```

File: lib/informativeTests/informativeTest_6_3_9.js [hunk 2]
```diff
@@ -94,7 +94,7 @@ export default function informativeTest_6_3_9(doc) {
                 message:
                   'order of ancestors with categories vendor, product_name, product_version is not correct',
               })
-              return
+              break
```
Comment (tschmidtb51) [inline lib/informativeTests/informativeTest_6_3_9.js:97]: :+1:

File: lib/informativeTests/informativeTest_6_3_9.js [hunk 3]
```diff
@@ -81,7 +81,7 @@ export default function informativeTest_6_3_9(doc) {
                 instancePath: `${prefix}${branchIndex}`,
                 message: `missing ancestor with category ${index.category}`,
               })
-              return
+              break
             }
           }
           const sortedIndexes = mandatoryCategoryIndexes
```
Comment (tschmidtb51) [inline lib/informativeTests/informativeTest_6_3_9.js:87]: Consequently, we need to filter those values, that are not present. The following code could work - but needs a bit polishing in terns of naming variables (esp. `presentIndexes`):
```
//remove missing categories
          const presentIndexes = mandatoryCategoryIndexes
            .filter((i) => i.index > -1)
          const sortedIndexes = presentIndexes
            .slice()
            .sort((a, z) => a.index - z.index)
          for (const [i, sortedIndex] of sortedIndexes.entries()) {
            if (sortedIndex.category !== presentIndexes[i].category) {

**Review by tschmidtb51 — CHANGES_REQUESTED** [review]
I had a look into it and found a few things. Please have a look at my comments.

This should result in the following information count:
- `oasis_csaf_tc-csaf_2_0-2021-6-3-09-01.json`: 1
- `oasis_csaf_tc-csaf_2_0-2021-6-3-09-02.json`: 2
- `oasis_csaf_tc-csaf_2_0-2021-6-3-09-03.json`: 4
- `oasis_csaf_tc-csaf_2_0-2021-6-3-09-04.json`: 2
- `oasis_csaf_tc-csaf_2_0-2021-6-3-09-05.json`: 2
- `oasis_csaf_tc-csaf_2_0-2021-6-3-09-06.json`: 6

**Review by tschmidtb51 — COMMENTED** [review]
Implementation looks good. Please add https://github.com/oasis-tcs/csaf/blob/master/csaf_2.0/test/validator/data/informative/oasis_csaf_tc-csaf_2_0-2021-6-3-09-06.json to complete.

**Comment by domachine** [comment]
> Implementation looks good. Please add https://github.com/oasis-tcs/csaf/blob/master/csaf_2.0/test/validator/data/informative/oasis_csaf_tc-csaf_2_0-2021-6-3-09-06.json to complete.

Yes, missed that! Thank you.

**Review by tschmidtb51 — DISMISSED** [review]
LGTM

File: tests/informativeTest_6_3_9.js [hunk 4]
```diff
@@ -16,6 +16,19 @@ describe('Informative test 6.3.9', function () {
...
         expect(result.infos.length).to.be.greaterThan(0)
+
+        const expectedErrorCounts = new Map([
+          ['oasis_csaf_tc-csaf_2_0-2021-6-3-09-01.json', 1],
+          ['oasis_csaf_tc-csaf_2_0-2021-6-3-09-02.json', 2],
+          ['oasis_csaf_tc-csaf_2_0-2021-6-3-09-03.json', 4],
+          ['oasis_csaf_tc-csaf_2_0-2021-6-3-09-04.json', 2],
+          ['oasis_csaf_tc-csaf_2_0-2021-6-3-09-05.json', 2],
+          ['oasis_csaf_tc-csaf_2_0-2021-6-3-09-06.json', 6],
+        ])
```
Comment (valentindey) [inline tests/informativeTest_6_3_9.js:27]: couldn't this be moved out of the loop?

**Review by valentindey — APPROVED** [review]

File: tests/informativeTest_6_3_9.js (hunk 4 above)
Comment (tschmidtb51) [reply tests/informativeTest_6_3_9.js:27]: Good catch. I guess it could/should....

Comment (domachine) [reply tests/informativeTest_6_3_9.js]: Alright, I changed that and therefore need reviews again ;-)

**Review by valentindey — APPROVED** [review]

**Review by tschmidtb51 — APPROVED** [review]
LGTM

**Pull request merged**